File: 03-GildedRose/go/corevill/gilded-rose.go
package main

type Item struct {
	name            string
	sellIn, quality int
}

var items = []*Item{
	&Item{"+5 Dexterity Vest", 10, 20},
	&Item{"Aged Brie", 2, 0},
	&Item{"Elixir of the Mongoose", 5, 7},
	&Item{"Sulfuras, Hand of Ragnaros", 0, 80},
	&Item{"Conjured Mana Cake", 3, 6},
	&Item{"Backstage passes to a TAFKAL80ETC concert", 15, 20},
	&Item{"Generic Name", 15, 20},
}

const DEXTERITY = "+5 Dexterity Vest"
const AGED_BRIE = "Aged Brie"
const ELIXIR = "Elixir of the Mongoose"
const SULFURAS = "Sulfuras, Hand of Ragnaros"
const PASSES = "Backstage passes to a TAFKAL80ETC concert"
const CONJURED = "Conjured Mana Cake"
const GENERIC_ITEM = "Generic Name"

const QUALITY_UNIT = 1
const SELLIN_UNIT = 1

const MAX_QUALITY = 50
const MIN_QUALITY = 0

const SELLIN_EXPIRED_VALUE = 0

const PASSES_THRESHOLD_DOUBLE_QUALITY = 10
const PASSES_THRESHOLD_TRIPLE_QUALITY = 5

func main() {
	updateQuality()
}

/* NAMES COMPARE */
func isNotAged(itemName string) bool {
	return itemName != AGED_BRIE
}

func isNotPasses(itemName string) bool {
	return itemName != PASSES
}

func isNotSulfuras(itemName string) bool {
	return itemName != SULFURAS
}

/* QUALITY */
func notMaxQuality(itemQuality int) bool {
	return itemQuality < MAX_QUALITY
}

func notMinQuality(itemQuality int) bool {
	return itemQuality > MIN_QUALITY
}

/* ADD AND SUBSTRACT QUALITY */
func increaseQuality(item *Item) {
	item.quality += QUALITY_UNIT
}

func decreaseQuality(item *Item) {
	item.quality -= QUALITY_UNIT
}

/* ADD AND SUBSTRACT SELLIN */
func decreaseSellIn(item *Item) {
	item.sellIn -= SELLIN_UNIT
}

func isExpired(item *Item) bool {
	return item.sellIn < SELLIN_EXPIRED_VALUE
}

func setZeroQuality(item *Item) {
	item.quality = 0
}

func updateQuality() {
	for i := 0; i < len(items); i++ {
		item := items[i]
		itemName := item.name
		itemQuality := item.quality

		isNotAged := isNotAged(itemName)
		isNotPasses := isNotPasses(itemName)
		isPasses := !isNotPasses
		isNotSulfuras := isNotSulfuras(itemName)

		belowThreasholdDoubleQuality := item.sellIn <= PASSES_THRESHOLD_DOUBLE_QUALITY
		belowThreasholdTripleQuality := item.sellIn <= PASSES_THRESHOLD_TRIPLE_QUALITY

		if isNotAged && isNotPasses {
			if notMinQuality(itemQuality) {
				if isNotSulfuras {
					decreaseQuality(items[i])
				}
			}
		} else {
			if notMaxQuality(itemQuality) {
				increaseQuality(items[i])
				if isPasses {
					if belowThreasholdDoubleQuality {
						if notMaxQuality(itemQuality) {
							increaseQuality(items[i])
						}
					}
					if belowThreasholdTripleQuality {
						if notMaxQuality(itemQuality) {
							increaseQuality(items[i])
						}
					}
				}
			}
		}

		if isNotSulfuras {
			decreaseSellIn(items[i])
		}

		if isExpired(items[i]) {
			if isNotAged {
				if isNotPasses {
					if notMinQuality(itemQuality) {
						if isNotSulfuras {
							decreaseQuality(items[i])
						}
					}
				} else {
					setZeroQuality(items[i])
				}
			} else {
				if notMaxQuality(itemQuality) {
					increaseQuality(items[i])
				}
			}
		}
	}
}

File: 03-GildedRose/java/group1part1/src/main/java/com/ccsw/coedevon/codingdojo/gildedrose/GildedRose.java
package com.ccsw.coedevon.codingdojo.gildedrose;

import java.util.List;

public class GildedRose {

  /**
   *
   */
  private static final int DEADLINE_SELL_IN = 0;

  /**
   *
   */
  private static final int SELL_IN_UNIT = 1;

  /**
   *
   */
  private static final int PASSES_THRESHOLD_TRIPLE_QUALITY = 5;

  /**
   *
   */
  private static final int PASSES_THRESHOLD_DOUBLE_QUALITY = 10;

  /**
   *
   */
  private static final int MAX_QUALITY = 50;

  /**
   *
   */
  private static final int QUALITY_UNIT = 1;

  /**
   *
   */
  private static final String SULFURAS = "Sulfuras, Hand of Ragnaros";

  /**
   *
   */
  private static final int MIN_QUALITY = 0;

  /**
   *
   */
  private static final String BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert";

  /**
   *
   */
  private static final String AGED_BRIE = "Aged Brie";

  private List<Item> items;

  public GildedRose(List<Item> items) {
    this.items = items;
  }

  public void updateQuality() {

    for (Item item : this.items) {
      String itemName = item.getName();

      boolean isNotAgedBrie = !AGED_BRIE.equals(itemName);
      boolean isPasses = BACKSTAGE_PASSES.equals(itemName);
      boolean isNotPasses = !isPasses;
      boolean isNotSulfuras = !SULFURAS.equals(itemName);

      boolean belowThresholdDoubleQuality = item.getSellIn() <= PASSES_THRESHOLD_DOUBLE_QUALITY;
      boolean belowThresholdTripleQuality = item.getSellIn() <= PASSES_THRESHOLD_TRIPLE_QUALITY;

      if (isNotAgedBrie && isNotPasses) {
        if (notMinQuality(item)) {
          if (isNotSulfuras) {
            decreaseQuality(item);
          }
        }
      } else {
        if (notMaxQuality(item)) {
          increaseQuality(item);

          if (isPasses) {
            if (belowThresholdDoubleQuality) {
              if (notMaxQuality(item)) {
                increaseQuality(item);
              }
            }

            if (belowThresholdTripleQuality) {
              if (notMaxQuality(item)) {
                increaseQuality(item);
              }
            }
          }
        }
      }

      if (isNotSulfuras) {
        decreaseSellIn(item);
      }

      if (hasExpired(item)) {
        if (isNotAgedBrie) {
          if (isNotPasses) {
            if (notMinQuality(item)) {
              if (isNotSulfuras) {
                decreaseQuality(item);
              }
            }
          } else {
            setZeroQuality(item);
          }
        } else {
          if (notMaxQuality(item)) {
            increaseQuality(item);
          }
        }
      }
    }
  }

  /**
   * @param item
   */
  private void setZeroQuality(Item item) {

    item.setQuality(0);
  }

  /**
   * @param item
   * @return
   */
  private boolean hasExpired(Item item) {

    boolean hasExpired = item.getSellIn() < DEADLINE_SELL_IN;
    return hasExpired;
  }

  /**
   * @param item
   */
  private void decreaseSellIn(Item item) {

    item.setSellIn(item.getSellIn() - SELL_IN_UNIT);
  }

  /**
   * @param item
   * @return
   */
  private boolean notMaxQuality(Item item) {

    boolean notMaxQuality = item.getQuality() < MAX_QUALITY;
    return notMaxQuality;
  }

  /**
   * @param item
   * @return
   */
  private boolean notMinQuality(Item item) {

    boolean notMinQuality = item.getQuality() > MIN_QUALITY;
    return notMinQuality;
  }

  /**
   * @param item
   */
  private void increaseQuality(Item item) {

    item.setQuality(item.getQuality() + QUALITY_UNIT);
  }

  /**
   * @param item
   */
  private void decreaseQuality(Item item) {

    item.setQuality(item.getQuality() - QUALITY_UNIT);
  }
}